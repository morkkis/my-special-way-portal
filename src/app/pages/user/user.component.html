<div class="main-content">
  <div class="container-fluid">
    <div class="header-menu-container">
      <div class="example-header">
        <button mat-raised-button color="primary" (click)="addNewUser()">הוספת משתמש חדש</button>
      </div>
      <div class="example-header">
        <mat-form-field class="search-user">
          <input matInput (keyup)="applyFilter($event.target.value)" placeholder="פילטר">
        </mat-form-field>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="example-table"
                   matSort matSortActive="firstName" matSortDisableClear matSortDirection="asc">

          <!-- ID Column -->
          <ng-container matColumnDef="name">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם</mat-header-cell>
            <mat-cell *matCellDef="let row"> {{row.firstName}} {{ row.lastName }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="userName">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row"> {{row.userName}}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="class">
            <mat-header-cell *matHeaderCellDef mat-sort-header>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row"> {{row._class}}</mat-cell>
          </ng-container>

          <!-- Name Column -->
          <ng-container matColumnDef="type">
            <mat-header-cell *matHeaderCellDef mat-sort-header>סוג היוזר</mat-header-cell>
            <mat-cell *matCellDef="let row">{{toHebrew(row.userType)}}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="editDetails">
            <mat-header-cell *matHeaderCellDef>עריכת פרטים</mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" (click)="editUser(row)">ערוך</button>
            </mat-cell>
          </ng-container>

          <!--&lt;!&ndash; State Column &ndash;&gt;-->
          <ng-container matColumnDef="deleteUser">-->
            <mat-header-cell *matHeaderCellDef>מחיקת משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" (click)="deleteUser(row.id)">מחק</button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;">
        </mat-row>
        </mat-table>

        <mat-paginator [pageSizeOptions]="[5, 10, 25, 100]"></mat-paginator>
      </div>
    </div>
  </div>
</div>
