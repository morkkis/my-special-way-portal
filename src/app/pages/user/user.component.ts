import { Component, ElementRef, OnInit, ViewChild, AfterViewInit } from '@angular/core';
import { MatDialog, MatPaginator, MatSort, MatTableDataSource } from '@angular/material';
import { UserService } from './services/user.service';
import { merge } from 'rxjs/observable/merge';
import { of as observableOf } from 'rxjs/observable/of';
import { catchError } from 'rxjs/operators/catchError';
import { map } from 'rxjs/operators/map';
import { startWith } from 'rxjs/operators/startWith';
import { switchMap } from 'rxjs/operators/switchMap';
import { User, UserType } from '../../models/user.model';
import { AddUserDialogComponent } from './dialogs/add/add-user.dialog';

@Component({
  selector: 'app-user',
  templateUrl: './user.component.html',
  styleUrls: ['./user.component.scss'],
})
export class UserComponent implements OnInit, AfterViewInit {

  displayedColumns = ['name', 'userName', 'class', 'type', 'editDetails', 'deleteUser'];
  dataSource = new MatTableDataSource();
  resultsLength = 0;

  @ViewChild(MatPaginator) paginator: MatPaginator;
  @ViewChild(MatSort) sort: MatSort;
  @ViewChild('table') table: ElementRef;

  constructor(
    private userService: UserService,
    public dialog: MatDialog,
  ) { }

  ngOnInit(): void {

    this.sort.sortChange.subscribe(() => this.paginator.pageIndex = 0);
    merge(this.sort.sortChange, this.paginator.page)
      .pipe(
        startWith({}),
        switchMap(() => {
          if (this.userService !== undefined) {
            return this.userService.getAllUsers();
          }
        }),
        map((data) => {
          return data.data.allUsers;
        }),
        catchError(() => {
          return observableOf([]);
        }),
    ).subscribe((data) => this.dataSource.data = data);

  }

  ngAfterViewInit() {
    this.dataSource.paginator = this.paginator;
    this.dataSource.sort = this.sort;
  }
  applyFilter(filterValue: string) {
    filterValue = filterValue.trim(); // Remove whitespace
    filterValue = filterValue.toLowerCase(); // Datasource defaults to lowercase matches
    this.dataSource.filter = filterValue;
  }

  toHebrew(type: UserType) {
    return UserType[type];
  }
  addNewUser() {
    const dialogRef = this.dialog.open(AddUserDialogComponent, {
      height: '600px',
      width: '350px',
    });
    dialogRef.afterClosed().subscribe((data) => {
      if (data) {
        // After dialog is closed we're doing frontend updates
        // For add we're just pushing a new row inside DataService
        this.dataSource.data.push(data);
        this.dataSource.paginator = this.paginator;
        this.userService.create(data);
      }
    });
  }

  deleteUser(id: number) {
    this.userService.delete(id);
  }

  editUser(user: User) {
    this.userService.update(user);
  }

}
