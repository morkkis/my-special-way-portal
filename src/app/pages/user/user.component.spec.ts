jest.mock('./services/user.service');

import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRowDef, MatRowDef, MatHeaderRow, MatDialog } from '@angular/material';
import { UserComponent } from './user.component';
import { UserService } from './services/user.service';

describe('dashboard component', () => {
  beforeEach(async () => {

    class UserDialogMock {
      open = jest.fn().mockImplementation(() => {
        return {
          afterClosed: () => {
            return {
              subscribe: jest.fn(),
            };
          },
        };
      });
    }

    TestBed.configureTestingModule({
      imports: [

      ],
      declarations: [
        UserComponent,
        MatHeaderRow,
        MatRowDef,
        MatHeaderRowDef,
      ],
      providers: [
        UserService,
        { provide: MatDialog, useClass: UserDialogMock },

      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

    // shallow = new Shallow(UserComponent, AppModule);
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(UserComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should open dialog when calling addNewUser function', () => {
    const fixture = TestBed.createComponent(UserComponent);
    fixture.componentInstance.addNewUser();

    const userDialogMock = TestBed.get(MatDialog);
    expect(userDialogMock.open).toHaveBeenCalled();
  });
});
