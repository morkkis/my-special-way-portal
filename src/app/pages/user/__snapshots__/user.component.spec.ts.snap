// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard component should render component as described in snapshot 1`] = `
<app-user
  dataSource={[Function MatTableDataSource]}
  dialog={[Function UserDialogMock]}
  displayedColumns={[Function Array]}
  paginator="undefined"
  resultsLength="0"
  sort="undefined"
  table={[Function ElementRef]}
  userService={[Function UserService]}
>
  <div
    class="main-content"
  >
    
    
    <div
      class="container-fluid"
    >
      
      
      <div
        class="header-menu-container"
      >
        
        
        <div
          class="example-header"
        >
          
          
          <button
            color="primary"
            mat-raised-button=""
          >
            הוספת משתמש חדש
          </button>
          
        
        </div>
        
        
        <div
          class="example-header"
        >
          
          
          <mat-form-field
            class="search-user"
          >
            
            
            <input
              matinput=""
              placeholder="פילטר"
            />
            
          
          </mat-form-field>
          
        
        </div>
        
      
      </div>
      
      
      <div
        class="row"
      >
        
        
        <div
          class="col-lg-12 col-md-12 col-sm-12"
        >
          
          
          <mat-table
            class="example-table"
            matsort=""
            matsortactive="firstName"
            matsortdirection="asc"
            matsortdisableclear=""
          >
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            --&gt;
              
            
            
              
            
            
            
            
  
            
            
            
            
            
            
          
          </mat-table>
          
  
          
          <mat-paginator />
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  
</app-user>
`;
